### ISSUE TITLE
BuildWidthTable doesn't build on old GHC

### ISSUE THREAD

**Issue by glguy** [jtdaugherty/vty#185]
I got this build failure on an old 7.10 build. We should either update the base lowerbound or make it build again on old GHC

```
In-place registering vty-5.28.1...
Preprocessing executable 'vty-build-width-table' for vty-5.28.1...
tools/BuildWidthTable.hs:9:8:
    Could not find module ‘Data.Semigroup’
    It is a member of the hidden package ‘semigroups-0.19.1@semig_BSdH1CQRYnR6ibErfUHmLL’.
    Perhaps you need to add ‘semigroups’ to the build-depends in your .cabal file.
    Use -v to see a list of the files searched for.
Failed to install vty-5.28.1
cabal: Error: some packages failed to install:
vty-5.28.1 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by jtdaugherty** [comment]
It looks like just adding this to the `executable` section is enough, but I don't know if it's the right way to handle this:
```
   if !impl(ghc >= 8.0)
     build-depends:     semigroups >= 0.16
```
(That is just borrowed from the similar `build-depends` configuration for the library itself.)

**Comment by jtdaugherty** [comment]
Granted, even with that resolved, there are other warnings on 7.10.3 that I'm seeing:
```
In file included from /var/folders/wd/l02zlzbd62538wwybvjhv33w0000gq/T/ghc7586_0/ghc_3.c:1:0:


In file included from /Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/Rts.h:217:0:


/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:505:5:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF || ZERO_SLOP_FOR_SANITY_CHECK
    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:502:37:
     note: expanded from macro 'ZERO_SLOP_FOR_LDV_PROF'
#define ZERO_SLOP_FOR_LDV_PROF     (defined(PROFILING))
                                    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:505:31:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF || ZERO_SLOP_FOR_SANITY_CHECK
                              ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:503:37:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:505:31:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:503:56:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                                       ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:523:5:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:502:37:
     note: expanded from macro 'ZERO_SLOP_FOR_LDV_PROF'
#define ZERO_SLOP_FOR_LDV_PROF     (defined(PROFILING))
                                    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:523:32:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
                               ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:503:37:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:523:32:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:503:56:
     note: expanded from macro 'ZERO_SLOP_FOR_SANITY_CHECK'
#define ZERO_SLOP_FOR_SANITY_CHECK (defined(DEBUG) && !defined(THREADED_RTS))
                                                       ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:552:5:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:502:37:
     note: expanded from macro 'ZERO_SLOP_FOR_LDV_PROF'
#define ZERO_SLOP_FOR_LDV_PROF     (defined(PROFILING))
                                    ^

/Users/cygnus/.ghcup/ghc/7.10.3/lib/ghc-7.10.3/include/rts/storage/ClosureMacros.h:552:32:
     warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ZERO_SLOP_FOR_LDV_PROF && !ZERO_SLOP_FOR_SANITY_CHECK
                               ^
```
And so on.

**Comment by glguy** [comment]
This warning suggests to me that the version of GHC isn't that compatible with the OS you're building on, rather than that vty is wrong.

**Comment by jtdaugherty** [comment]
Strange; I just installed it with `ghcup install 7.10.3`, but probably it just has issues with macOS Mojave. So assuming I can ignore those warnings, what do you think of the fix I posted above?

**Comment by glguy** [comment]
I think your fix is the right approach. Sorry I didn't put more effort into a fix yesterday. I had something else going on and just made the issue to ensure this didn't get dropped.

**Comment by jtdaugherty** [comment]
No problem! Thanks!

**Comment by jtdaugherty** [comment]
Released in `5.28.2`.
